File: MarkdownEditor/MarkdownEditor/Models/GitStatus.swift
import Foundation
import CryptoKit

// MARK: - Repository Snapshot

/// Complete snapshot of a repo's state, replacing RepoMetadata.
struct RepoSnapshot: Codable {
    let fullName: String
    let cloneURL: String
    let defaultBranch: String
    var lastSyncedAt: String
    var headCommitSHA: String
    var treeSHA: String
    var files: [String: TrackedFile]

    /// Create from legacy RepoMetadata (backward compat)
    init(from metadata: RepoMetadata) {
        self.fullName = metadata.fullName
        self.cloneURL = metadata.cloneURL
        self.defaultBranch = metadata.defaultBranch
        self.lastSyncedAt = metadata.lastSyncedAt
        self.headCommitSHA = ""
        self.treeSHA = ""
        self.files = [:]
    }

    init(fullName: String, cloneURL: String, defaultBranch: String,
         lastSyncedAt: String, headCommitSHA: String, treeSHA: String,
         files: [String: TrackedFile]) {
        self.fullName = fullName
        self.cloneURL = cloneURL
        self.defaultBranch = defaultBranch
        self.lastSyncedAt = lastSyncedAt
        self.headCommitSHA = headCommitSHA
        self.treeSHA = treeSHA
        self.files = files
    }
}

// MARK: - Tracked File

/// Per-file tracking info stored in the snapshot.
struct TrackedFile: Codable {
    let sha: String               // GitHub blob SHA
    let originalContentHash: String // SHA-256 of content at clone/pull time
}

// MARK: - File Change

/// Represents a local change detected by status().
struct FileChange: Identifiable {
    let id = UUID()
    let path: String
    let changeType: FileChangeType
    var isSelected: Bool = true
    var additions: Int = 0
    var deletions: Int = 0
}

enum FileChangeType: String, Codable {
    case modified = "Modified"
    case added = "Added"
    case deleted = "Deleted"
}

// MARK: - Sync State

enum SyncState: Equatable {
    case unknown
    case checking
    case upToDate
    case localChanges(count: Int)
    case remoteChanges
    case conflict
    case error(String)

    static func == (lhs: SyncState, rhs: SyncState) -> Bool {
        switch (lhs, rhs) {
        case (.unknown, .unknown), (.checking, .checking),
             (.upToDate, .upToDate), (.remoteChanges, .remoteChanges),
             (.conflict, .conflict):
            return true
        case (.localChanges(let a), .localChanges(let b)):
            return a == b
        case (.error(let a), .error(let b)):
            return a == b
        default:
            return false
        }
    }
}

// MARK: - Diff Models

struct FileDiff {
    let originalPath: String
    let modifiedPath: String
    let hunks: [DiffHunk]
    let additions: Int
    let deletions: Int
}

struct DiffHunk {
    let oldStart: Int
    let oldCount: Int
    let newStart: Int
    let newCount: Int
    let lines: [DiffLine]

    var header: String {
        "@@ -\(oldStart),\(oldCount) +\(newStart),\(newCount) @@"
    }
}

struct DiffLine: Identifiable {
    let id = UUID()
    let type: DiffLineType
    let content: String
    let oldLineNumber: Int?
    let newLineNumber: Int?
}

enum DiffLineType {
    case context
    case addition
    case deletion
}

// MARK: - Pull Result

enum PullResult {
    case upToDate
    case updated(filesChanged: Int)
    case conflicts(files: [String])
}

// MARK: - Utility

enum ContentHasher {
    static func sha256(_ string: String) -> String {
        let data = Data(string.utf8)
        let hash = SHA256.hash(data: data)
        return hash.compactMap { String(format: "%02x", $0) }.joined()
    }

    static func sha256(data: Data) -> String {
        let hash = SHA256.hash(data: data)
        return hash.compactMap { String(format: "%02x", $0) }.joined()
    }
}

File: android/lib/models/git_status.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

// MARK: - Repository Snapshot

class RepoSnapshot {
  final String fullName;
  final String cloneUrl;
  final String defaultBranch;
  String lastSyncedAt;
  String headCommitSha;
  String treeSha;
  Map<String, TrackedFile> files;

  RepoSnapshot({
    required this.fullName,
    required this.cloneUrl,
    required this.defaultBranch,
    required this.lastSyncedAt,
    required this.headCommitSha,
    required this.treeSha,
    required this.files,
  });

  factory RepoSnapshot.fromJson(Map<String, dynamic> json) => RepoSnapshot(
        fullName: json['fullName'] as String,
        cloneUrl: json['cloneURL'] as String,
        defaultBranch: json['defaultBranch'] as String,
        lastSyncedAt: json['lastSyncedAt'] as String,
        headCommitSha: json['headCommitSHA'] as String? ?? '',
        treeSha: json['treeSHA'] as String? ?? '',
        files: (json['files'] as Map<String, dynamic>? ?? {}).map(
          (k, v) => MapEntry(k, TrackedFile.fromJson(v as Map<String, dynamic>)),
        ),
      );

  Map<String, dynamic> toJson() => {
        'fullName': fullName,
        'cloneURL': cloneUrl,
        'defaultBranch': defaultBranch,
        'lastSyncedAt': lastSyncedAt,
        'headCommitSHA': headCommitSha,
        'treeSHA': treeSha,
        'files': files.map((k, v) => MapEntry(k, v.toJson())),
      };
}

// MARK: - Tracked File

class TrackedFile {
  final String sha;
  final String originalContentHash;

  const TrackedFile({required this.sha, required this.originalContentHash});

  factory TrackedFile.fromJson(Map<String, dynamic> json) => TrackedFile(
        sha: json['sha'] as String,
        originalContentHash: json['originalContentHash'] as String,
      );

  Map<String, dynamic> toJson() => {
        'sha': sha,
        'originalContentHash': originalContentHash,
      };
}

// MARK: - File Change

enum FileChangeType { modified, added, deleted }

class FileChange {
  final String path;
  final FileChangeType changeType;
  bool isSelected;
  int additions;
  int deletions;

  FileChange({
    required this.path,
    required this.changeType,
    this.isSelected = true,
    this.additions = 0,
    this.deletions = 0,
  });
}

// MARK: - Sync State

sealed class SyncState {
  const SyncState();
}

class SyncStateUnknown extends SyncState {
  const SyncStateUnknown();
}

class SyncStateChecking extends SyncState {
  const SyncStateChecking();
}

class SyncStateUpToDate extends SyncState {
  const SyncStateUpToDate();
}

class SyncStateLocalChanges extends SyncState {
  final int count;
  const SyncStateLocalChanges(this.count);
}

class SyncStateRemoteChanges extends SyncState {
  const SyncStateRemoteChanges();
}

class SyncStateConflict extends SyncState {
  const SyncStateConflict();
}

class SyncStateError extends SyncState {
  final String message;
  const SyncStateError(this.message);
}

// MARK: - Diff Models

class FileDiff {
  final String originalPath;
  final String modifiedPath;
  final List<DiffHunk> hunks;
  final int additions;
  final int deletions;

  const FileDiff({
    required this.originalPath,
    required this.modifiedPath,
    required this.hunks,
    required this.additions,
    required this.deletions,
  });
}

class DiffHunk {
  final int oldStart;
  final int oldCount;
  final int newStart;
  final int newCount;
  final List<DiffLine> lines;

  const DiffHunk({
    required this.oldStart,
    required this.oldCount,
    required this.newStart,
    required this.newCount,
    required this.lines,
  });

  String get header => '@@ -$oldStart,$oldCount +$newStart,$newCount @@';
}

enum DiffLineType { context, addition, deletion }

class DiffLine {
  final DiffLineType type;
  final String content;
  final int? oldLineNumber;
  final int? newLineNumber;

  const DiffLine({
    required this.type,
    required this.content,
    this.oldLineNumber,
    this.newLineNumber,
  });
}

// MARK: - Pull Result

sealed class PullResult {
  const PullResult();
}

class PullResultUpToDate extends PullResult {
  const PullResultUpToDate();
}

class PullResultUpdated extends PullResult {
  final int filesChanged;
  const PullResultUpdated(this.filesChanged);
}

class PullResultConflicts extends PullResult {
  final List<String> files;
  const PullResultConflicts(this.files);
}

// MARK: - Content Hasher

class ContentHasher {
  static String sha256String(String text) {
    final bytes = utf8.encode(text);
    return sha256.convert(bytes).toString();
  }

  static String sha256Bytes(List<int> data) {
    return sha256.convert(data).toString();
  }
}
